### DISCUSSION TITLE
RegisterPluginStep Usage Help

### DISCUSSION THREAD

**Discussion by RobNightingale-LSL** [jordimontana82/fake-xrm-easy#556] [Q&A] [answered]
I have written a plugin to interrupt the PickFromQueue message. This has been tested manually and is working really well. I am struggling replicating the tests using FakeXrmEasy.
I am able to execute the PickFromQueueRequest and the queueitem is deleted. If the interrupt plugin worked then the queueitem should not be deleted (removed from the queue).
 
The line `fakedContext.RegisterPluginStep<InterruptPick>("PickFromQueue", ProcessingStepStage.Preoperation);` is not causing the plugin InterruptPick to be called when I debug and step throught the code.

Hopefully I am missing something obvious to get this test to trigger the plugin.

Full function: Create a queueitem and user then call PickFromQueue request and check that the queueitem has not been deleted.
```
[Fact]
public void InterruptCasePick_RemoveFromQueue_Success()
{
    var entities = new List<Entity>();
    var fakedContext = new XrmFakedContext { UsePipelineSimulation = true };
    
    var queueItem = new Entity("queueitem", Guid.NewGuid());
    queueItem.AddAttribute("createdon", DateTime.UtcNow);
    queueItem.AddAttribute("queueid", Guid.NewGuid());
    entities.Add(queueItem);

    var user = new Entity("systemuser", Guid.NewGuid());
    user.AddAttribute("createdon", DateTime.UtcNow);
    entities.Add(user);

    fakedContext.Initialize(entities);
    fakedContext.RegisterPluginStep<InterruptPick>("PickFromQueue", ProcessingStepStage.Preoperation);
    
    PickFromQueueRequest request = new PickFromQueueRequest();
    request.QueueItemId = queueItem.Id;
    request.RemoveQueueItem = true;
    request.WorkerId = user.Id;
    
    fakedContext.GetOrganizationService().Execute(request);
    
    var updatedQueueItem = fakedContext.GetFakedOrganizationService()
        .Retrieve("queueitem", queueItem.Id, new ColumnSet(true));

    Assert.True(updatedQueueItem.GetAttributeValue<OptionSetValue>("statecode").Value == 0);
}
```

**Comment by jordimontana82** [comment, marked as answer]
Hi Rob,

You're absolutely right. Right now v1 has a limitation where pipeline simulation is limited to trigger off CRUD messages only, not custom ones (OOB messages, custom actions / apis). 

However, this has been effectively rewritten in v2 to be able to stack multiple layers of executions into the one message, effectively enabling this and many other scenarios (this is the new middleware feature).

**Reply by jordimontana82** [reply]
https://github.com/DynamicsValue/announcements

**Reply by RobNightingale-LSL** [reply]
Thanks for clearing that up Jordi!

With that in mind do you have any thoughts how I could test that my interrupt is working in isolation? 
My original test looks like this, it does not delete the queueitem as in the original test above so the current assertion is not valid in this test.:

```
[Fact]
public void InterruptCasePick_RemoveFromQueue_Success()
{
    var entities = new List<Entity>();
    var fakedContext = new XrmFakedContext();
    
    var queueItem = new Entity("queueitem", Guid.NewGuid());
    queueItem.AddAttribute("createdon", DateTime.UtcNow);
    queueItem.AddAttribute("queueid", Guid.NewGuid());
    entities.Add(queueItem);

    var user = new Entity("systemuser", Guid.NewGuid());
    user.AddAttribute("createdon", DateTime.UtcNow);
    entities.Add(user);

    fakedContext.Initialize(entities);

    ParameterCollection inputParameters = new ParameterCollection();
    inputParameters.Add("QueueItemId", queueItem.Id);
    inputParameters.Add("RemoveQueueItem", true);
    inputParameters.Add("WorkerId", Guid.NewGuid());

    var fpec = new XrmFakedPluginExecutionContext()
    {
        MessageName = "PickFromQueue",
        Stage = 20,
        InputParameters = inputParameters,
        PostEntityImages = null
    };

    var response = fakedContext.ExecutePluginWith<InterruptPick>(fpec);

    var updatedQueueItem = fakedContext.GetOrganizationService()
        .Retrieve("queueitem", queueItem.Id, new ColumnSet(true));

    Assert.True(updatedQueueItem.GetAttributeValue<OptionSetValue>("statecode").Value == 0);
}
```

Essentially all the plugin does is change the InputParameter `RemoveQueueItem` from true to false. Do you know if there is a way to interrogate the response to check the maniluplation of the input parameters now that I cannot trust the assertion to check the deletion of the queueitem?

**Reply by jordimontana82** [reply]
Could you try this?

``` csharp
Assert.False(inputParameters["RemoveQueueItem"]);
```

**Reply by RobNightingale-LSL** [reply]
Well I feel silly now!!! That was one of the first things I tried in the watches but I think it might have been when I commented out the code within the plugin. Just tried again and `Assert.False((bool)inputParameters["RemoveQueueItem"]);` works a treat! 

Cheers Jordi!
